### PR TITLE
Add CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by billyb2**
### Change Summary

What and Why:
People new to the flyctl repo struggle to get around sometimes, so adding a file that explains basic things like building and runnning flyctl, as well as testing, creating a new release, etc. is something useful to share

How:
Create a CONTRIBUTING.md file with some relevant info on contributing to flyctl.

Related to:

---

### Documentation

- [ ] Fresh Produce
- [ ] In superfly/docs, or asked for help from docs team
- [x] n/a

### CODE REVIEW

**Comment by billyb2** [comment]
sorry michael and allison if yall got pinged. I ~stole~ borrowed some words yall wrote elsewhere and copied them in.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
+
+
+# Development flow
+
+
+## Building
+
+To build `flyctl`, all you need to do is to run `make build` from the `flyctl/` directory. This will build a binary in the `bin/` directory. Alternatively, you can run `go build .`
```
[the remark this answers is not in the document]
Comment (billyb2) [reply CONTRIBUTING.md:8]: I should clarify, yea

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+To run `flyctl`, you can run `go run .`, followed by whatever sub-command you want to run. So for example, to update a machine, you can run `go run . m update -a <app_name> <machine_id>`
+
+
+## Testing
+
+We have two different kinds of tests in `flyctl`, unit tests and integration tests (preflight). It&rsquo;s recommend to write a test for any features added or bug fixes, in order to prevent regressions in the future.
```
[the remark this answers is not in the document]
Comment (billyb2) [reply CONTRIBUTING.md:15]: org-mode export 😭

File: README.md [hunk 3]
```diff
@@ -98,42 +98,5 @@ There is a simple Powershell script, `winbuild.ps1`, which will run the code gen
...
-That builds a flyctl binary (just like running `make`), then runs the preflight tests against that binary.
-
-To run a single test:
-
-```
-make preflight-test T=TestAppsV2Example
-```
-
-Oh, add more preflight tests at `tests/preflight/*`
+## Development/Contributing guide
```
[the remark this answers is not in the document]
Comment (billyb2) [reply README.md:101]: Yea probably less confusing, I'll switch that to Contributing

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+`scripts/bump_version.sh`
+
+
+# Committing to flyctl
+
+When committing to `flyctl`, there are a few important things to keep in mind:
+
+-   Keep commits small and focused, it helps greatly when reviewing larger PRs
+-   Make sure to use descriptive messages in your commit messages, since we use those to generate changelogs. It also helps future people understand why a change was made.
```
[the remark this answers is not in the document]
Comment (billyb2) [reply CONTRIBUTING.md:73]: not sure what i meant here honestly, i'll fix it

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+When committing to `flyctl`, there are a few important things to keep in mind:
+
+-   Keep commits small and focused, it helps greatly when reviewing larger PRs
+-   Make sure to use descriptive messages in your commit messages, it helps future people understand why a change was made
+-   PRs are squash merged and used to generate changelogs, so please make sure to use descriptive titles
+
+
+## Examples
+
+[This is a bad example of a commit](https://github.com/superfly/flyctl/pull/1809/commits/6f167c858dbd7ae1324632dda9e29072ddde8ad7), it has a large diff and no explanation as to why this change is being made. [This is a great one](https://github.com/superfly/flyctl/commit/2636f47fe91cbe37018926cb0d7d2227a6887086), since it's a small commit, and it's reasoning as well as the context behind the change. Good commit messages also help contributors in the future to understand *why* we did something a certain way.
```
[the remark this answers is not in the document]
Comment (billyb2) [reply CONTRIBUTING.md:77]: my English is rough today

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
+# Development flow
+
+
+## Building
+
+To build `flyctl`, all you need to do is to run `make build` from the root directory. This will build a binary in the `bin/` directory. Alternatively, you can run `go build .`
+
+To run `flyctl`, you can just run the binary you built using `make build`: `./bin/flyctl`. So for example, to update a machine, you can run `go run . m update -a <app_name> <machine_id>`. Alternatively, you can build and run in the same command by running `go run .`, followed by whatever sub-command you want to run. Just note that this will have a slower startup.
```
[the remark this answers is not in the document]
Comment (billyb2) [reply CONTRIBUTING.md:8]: actually dumb oversight, thank you for catching this.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+To build `flyctl`, all you need to do is to run `make build` from the root directory. This will build a binary in the `bin/` directory. Alternatively, you can run `go build .`
+
+To run `flyctl`, you can just run the binary you built using `make build`: `./bin/flyctl`. So for example, to update a machine, you can run `./bin/flyctl m update -a <app_name> <machine_id>`. Alternatively, you can build and run in the same command by running `go run .`, followed by whatever sub-command you want to run. Just note that this will have a slower startup.
+
+
+## Testing
+
+We have two different kinds of tests in `flyctl`, unit tests and integration tests (preflight). It's recommended to write a test for any features added or bug fixes, in order to prevent regressions in the future.
+
```
Comment (michaeldwan) [inline CONTRIBUTING.md:14]: Can you put the command to run tests in a code block?

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+## Integration tests
+
+Unit tests are stored in individual files next to the functionality they're testing. For example
+
+`internal/command/secrets/parser_test.go`
+is a test for the secrets parsing code
+
+`internal/command/secrets/parser_test.go`.
+You can run these tests by running `make test` from the `flyctl/` directory.
```
Comment (michaeldwan) [inline CONTRIBUTING.md:32]: What is the `flyctl/` directory? root?

**Review by michaeldwan — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 8 above)
Comment (billyb2) [reply CONTRIBUTING.md:32]: yep :(. i'll fix it

File: CONTRIBUTING.md (hunk 7 above)
Comment (billyb2) [reply CONTRIBUTING.md:14]: do you mean like in the below sections regarding tests, or more than that?

Comment (michaeldwan) [reply CONTRIBUTING.md:14]: Oh just a command to run. People will be skimming this thinking "okay what do I run?" -- just need to provide some copy/pasteable commands to get them started

**Review by michaeldwan — APPROVED** [review]

**Pull request merged**